builtin/v13/migration/market.go

```diff
@@ -309,10 +309,21 @@ func (m *marketMigrator) migrateProviderSectorsAndStatesFromScratch(ctx context.
 		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load proposals array: %w", err)
 	}
 
+	proposalsSize := proposalsArr.Length()
+
 	var oldState market12.DealState
 	var newState market13.DealState
 
 	err = oldStateArray.ForEach(&oldState, func(i int64) error {
+		// Refresh proposals array periodically to avoid holding onto all ~10GB of memory
+		// throughout whole migration.
+		// This has limited impact on caching speedups because the access pattern is sequential.
+		if proposalsSize > 0 && i%(int64(proposalsSize)/1000) == 0 {
```

Comment:
Are we sure we don't have holes in this array? I think we want to count instead of using `i`.

Reply:
Thx, we will definitely have holes